src/util/battery/batterylinux.cpp

```diff
@@ -0,0 +1,88 @@
+#include "util/battery/batterylinux.h"
+
+// Contains 'signals' so we undefine Qt's 'signals'.
+#undef signals
+#include <upower.h>
+
+#include <QtDebug>
+
+BatteryLinux::BatteryLinux(QObject* pParent)
+    : Battery(pParent) {
+}
+
+BatteryLinux::~BatteryLinux() {
+}
+
+void BatteryLinux::read() {
+    m_iMinutesLeft = 0;
+    m_dPercentage = 0.0;
+    m_chargingState = Battery::UNKNOWN;
+
+    UpClient* client = up_client_new();
```

Comment:
Yea, I tried that but it doesn't actually work unless you create a new connection each time. up_client_get_devices returns nothing if you re-use the pointer.